byway/src/main/java/com/google/sps/servlets/EntityServlet.java

```diff
@@ -0,0 +1,115 @@
+        Trip trip = Trip.fromEntity(tripEntity);
+        userTrips.add(trip);
+      } catch (EntityNotFoundException exception) {
+        return;
+      }
+    }
+    String json = gson.toJson(userTrips);
+    response.setContentType("application/json");
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Key tripKey = createTripEntity();
+    String keyString = KeyFactory.keyToString(tripKey);
+    System.out.println(keyString);
+    addTripForUser(keyString);
+    String json = gson.toJson(KeyFactory.keyToString(tripKey));
+    response.setContentType("application/json");
+    response.getWriter().println(json);
+  }
+
+  /*
+   * Adds new Trip entity with empty properties
+   **/
+  public Key createTripEntity() {
+    System.setProperty(
+        DatastoreServiceConfig.DATASTORE_EMPTY_LIST_SUPPORT, Boolean.TRUE.toString());
+    Entity tripEntity = new Entity(Trip.DATASTORE_ENTITY_KIND);
+    tripEntity.setProperty("start", "");
+    tripEntity.setProperty("destinations", new ArrayList<String>());
+    tripEntity.setProperty("interests", new ArrayList<String>());
+    tripEntity.setProperty("route", new ArrayList<String>());
+    datastore.put(tripEntity);
+    Key tripKey = tripEntity.getKey();
+    tripEntity.setProperty("keyString", KeyFactory.keyToString(tripKey));
+    datastore.put(tripEntity);
+    return tripKey;
+  }
+
+  public UserInfo addUserEntity(com.google.appengine.api.users.User user) {
+    System.setProperty(
+        DatastoreServiceConfig.DATASTORE_EMPTY_LIST_SUPPORT, Boolean.TRUE.toString());
+    // Create a key based on the user ID
+    Key userKey = KeyFactory.createKey(UserInfo.DATASTORE_ENTITY_KIND, user.getUserId());
+    UserInfo userInfo;
+    try {
+      // try to retrieve the entity with the key
+      Entity userEntity = datastore.get(userKey);
+      datastore.put(userEntity);
+      userInfo = UserInfo.fromEntity(userEntity);
+    } catch (EntityNotFoundException exception) {
+      // If the user doesn't exist yet or is new, create a new user
+      Entity newUserEntity = new Entity(userKey);
+      newUserEntity.setProperty("email", user.getEmail());
+      newUserEntity.setProperty("userId", user.getUserId());
+      newUserEntity.setProperty("tripIds", new ArrayList<String>());
+      datastore.put(newUserEntity);
+      userInfo = UserInfo.fromEntity(newUserEntity);
+    }
+    return userInfo;
+  }
+
+  public void addTripForUser(String tripKey) {
+    Key userKey =
+        KeyFactory.createKey(
+            UserInfo.DATASTORE_ENTITY_KIND, userService.getCurrentUser().getUserId());
+    try {
+      Entity userEntity = datastore.get(userKey);
+      ArrayList<String> tripIds = (ArrayList<String>) userEntity.getProperty("tripIds");
+      tripIds.add(tripKey);
+      userEntity.setProperty("tripIds", tripIds);
+      datastore.put(userEntity);
```

Comment:
You may want to have a `toEntity` method on the `UserInfo` class. That way you can have something like

```java
userInfo.addTrip(tripKey);
datastore.put(userInfo.toEntity());
```

This allows you to only fetch the `UserInfo` object once, and re-use it later, and saves you from having to re-fetch the User entity.

Reply:
We are going to do this in another PR because Justine is currently working on it

Reply:
Actually I just added Justine's code to this PR, wasn't sure what to put in the catch block for the toEntity function though since the entity should already be created. Let me know if you have any ideas for this